File: python/src/pi_mono/coding_agent/modes/interactive/components/assistant_message.py
"""Assistant message rendering for interactive mode."""

from __future__ import annotations

from typing import Any

from pi_mono.agent.types import AgentMessage
from pi_mono.coding_agent.core.auth_guidance import format_api_error_message
from pi_mono.coding_agent.modes.interactive.theme.theme import get_markdown_theme, theme
from pi_mono.tui.components.markdown import DefaultTextStyle, Markdown
from pi_mono.tui.components.spacer import Spacer
from pi_mono.tui.components.text import Text
from pi_mono.tui.tui import Container

OSC133_ZONE_START = "\x1b]133;A\x07"
OSC133_ZONE_END = "\x1b]133;B\x07"
OSC133_ZONE_FINAL = "\x1b]133;C\x07"


class AssistantMessageComponent(Container):
    """Renders a complete assistant message with markdown and thinking blocks."""

    def __init__(
        self,
        message: AgentMessage | dict[str, Any] | None = None,
        *,
        hide_thinking_block: bool = False,
        hidden_thinking_label: str = "Thinking...",
    ) -> None:
        super().__init__()
        self._content_container = Container()
        self.add_child(self._content_container)
        self._hide_thinking_block = hide_thinking_block
        self._markdown_theme = get_markdown_theme()
        self._hidden_thinking_label = hidden_thinking_label
        self._last_message: AgentMessage | dict[str, Any] | None = None
        self._has_tool_calls = False
        if message is not None:
            self.update_content(message)

    def invalidate(self) -> None:
        super().invalidate()
        if self._last_message is not None:
            self.update_content(self._last_message)

    def set_hide_thinking_block(self, hide: bool) -> None:
        self._hide_thinking_block = hide
        if self._last_message is not None:
            self.update_content(self._last_message)

    def update_content(self, message: AgentMessage | dict[str, Any]) -> None:
        self._last_message = message
        self._content_container.clear()

        content = message.get("content", [])
        has_visible_content = any(
            (block.get("type") == "text" and str(block.get("text", "")).strip())
            or (block.get("type") == "thinking" and str(block.get("thinking", "")).strip())
            for block in content
        )

        if has_visible_content:
            self._content_container.add_child(Spacer(1))

        for index, block in enumerate(content):
            block_type = block.get("type")
            if block_type == "text":
                text = str(block.get("text", "")).strip()
                if text:
                    self._content_container.add_child(Markdown(text, 1, 0, self._markdown_theme))
            elif block_type == "thinking":
                thinking = str(block.get("thinking", "")).strip()
                if not thinking:
                    continue
                has_visible_after = any(
                    (later.get("type") == "text" and str(later.get("text", "")).strip())
                    or (later.get("type") == "thinking" and str(later.get("thinking", "")).strip())
                    for later in content[index + 1 :]
                )
                if self._hide_thinking_block:
                    self._content_container.add_child(
                        Text(
                            theme.italic(theme.fg("thinkingText", self._hidden_thinking_label)),
                            padding_x=1,
                            padding_y=0,
                        )
                    )
                else:
                    self._content_container.add_child(
                        Markdown(
                            thinking,
                            1,
                            0,
                            self._markdown_theme,
                            default_text_style=DefaultTextStyle(
                                color=theme.fg_fn("thinkingText"),
                                italic=True,
                            ),
                        )
                    )
                if has_visible_after:
                    self._content_container.add_child(Spacer(1))

        self._has_tool_calls = any(block.get("type") == "toolCall" for block in content)
        if not self._has_tool_calls:
            stop_reason = message.get("stopReason")
            if stop_reason == "aborted":
                error_message = message.get("errorMessage")
                if error_message and error_message != "Request was aborted":
                    abort_message = str(error_message)
                else:
                    abort_message = "Operation aborted"
                self._content_container.add_child(Spacer(1))
                self._content_container.add_child(
                    Text(theme.fg("error", abort_message), padding_x=1, padding_y=0)
                )
            elif stop_reason == "error":
                error_msg = format_api_error_message(
                    str(message.get("errorMessage") or "Unknown error"),
                    provider=str(message.get("provider") or ""),
                    model_id=str(message.get("model") or ""),
                )
                self._content_container.add_child(Spacer(1))
                self._content_container.add_child(
                    Text(theme.fg("error", error_msg), padding_x=1, padding_y=0)
                )

    def render(self, width: int) -> list[str]:
        lines = super().render(width)
        if self._has_tool_calls or not lines:
            return lines
        lines[0] = OSC133_ZONE_START + lines[0]
        lines[-1] = OSC133_ZONE_END + OSC133_ZONE_FINAL + lines[-1]
        return lines

File: packages/coding-agent/src/modes/interactive/components/assistant-message.ts
import type { AssistantMessage } from "@earendil-works/pi-ai";
import { Container, Markdown, type MarkdownTheme, Spacer, Text } from "@earendil-works/pi-tui";
import type { MarkdownTransformer } from "../../../core/extensions/types.ts";
import { getMarkdownTheme, theme } from "../theme/theme.ts";
import { createMarkdownTransform } from "./markdown-transform.ts";

const OSC133_ZONE_START = "\x1b]133;A\x07";
const OSC133_ZONE_END = "\x1b]133;B\x07";
const OSC133_ZONE_FINAL = "\x1b]133;C\x07";

/**
 * Component that renders a complete assistant message
 */
export class AssistantMessageComponent extends Container {
	private contentContainer: Container;
	private hideThinkingBlock: boolean;
	private markdownTheme: MarkdownTheme;
	private hiddenThinkingLabel: string;
	private outputPad: number;
	private markdownTransformers: readonly MarkdownTransformer[];
	private lastMessage?: AssistantMessage;
	private hasToolCalls = false;
	private isStreaming = false;

	// Text/thinking components created by the last content build, in content
	// order. During streaming, message_update events arrive per chunk; when the
	// run structure is unchanged the Markdown components are updated in place via
	// setText() instead of being destroyed and recreated, so their append-aware
	// render cache re-renders only the tail that grew.
	private streamingEditable: Array<{ kind: "text" | "thinking"; component: Markdown | Text }> = [];

	/** Read-only view for tests: components created by the last content build. */
	get streamingEditableComponents(): ReadonlyArray<{ kind: "text" | "thinking"; component: Markdown | Text }> {
		return this.streamingEditable;
	}

	private builtWithTheme?: MarkdownTheme;
	private builtWithOutputPad = 0;
	private builtWithHideThinkingBlock = false;
	private builtWithHiddenThinkingLabel = "";
	private builtWithTransformers: readonly MarkdownTransformer[] = [];

	constructor(
		message?: AssistantMessage,
		hideThinkingBlock = false,
		markdownTheme: MarkdownTheme = getMarkdownTheme(),
		hiddenThinkingLabel = "Thinking...",
		outputPad = 1,
		markdownTransformers: readonly MarkdownTransformer[] = [],
	) {
		super();

		this.hideThinkingBlock = hideThinkingBlock;
		this.markdownTheme = markdownTheme;
		this.hiddenThinkingLabel = hiddenThinkingLabel;
		this.outputPad = outputPad;
		this.markdownTransformers = markdownTransformers;

		// Container for text/thinking content
		this.contentContainer = new Container();
		this.addChild(this.contentContainer);

		if (message) {
			this.updateContent(message);
		}
	}

	override invalidate(): void {
		super.invalidate();
		if (this.lastMessage) {
			this.updateContent(this.lastMessage);
		}
	}

	setHideThinkingBlock(hide: boolean): void {
		this.hideThinkingBlock = hide;
		if (this.lastMessage) {
			this.updateContent(this.lastMessage);
		}
	}

	setHiddenThinkingLabel(label: string): void {
		this.hiddenThinkingLabel = label;
		if (this.lastMessage) {
			this.updateContent(this.lastMessage);
		}
	}

	setOutputPad(padding: number): void {
		this.outputPad = padding;
		if (this.lastMessage) {
			this.updateContent(this.lastMessage);
		}
	}

	override render(width: number): string[] {
		const lines = super.render(width);
		if (this.hasToolCalls || lines.length === 0) {
			return lines;
		}

		lines[0] = OSC133_ZONE_START + lines[0];
		lines[lines.length - 1] = OSC133_ZONE_END + OSC133_ZONE_FINAL + lines[lines.length - 1];
		return lines;
	}

	updateContent(message: AssistantMessage, isStreaming = this.isStreaming): void {
		const previousMessage = this.lastMessage;
		this.lastMessage = message;
		this.isStreaming = isStreaming;

		// Fast path: during streaming the content usually just grows. If the
		// sequence of text/thinking runs is unchanged and nothing that affects the
		// components' construction (theme, padding, ...) changed, update the
		// existing Markdown components in place so their append-aware render cache
		// stays warm.
		if (
			isStreaming &&
			previousMessage !== undefined &&
			previousMessage.stopReason === message.stopReason &&
			previousMessage.errorMessage === message.errorMessage &&
			this.tryUpdateStreamingInPlace(previousMessage, message)
		) {
			this.hasToolCalls = message.content.some((c) => c.type === "toolCall");
			return;
		}
		this.streamingEditable = [];

		// Clear content container
		this.contentContainer.clear();

		const hasVisibleContent = message.content.some(
			(c) => (c.type === "text" && c.text.trim()) || (c.type === "thinking" && c.thinking.trim()),
		);

		if (hasVisibleContent) {
			this.contentContainer.addChild(new Spacer(1));
		}

		// Render content in order
		const editable: Array<{ kind: "text" | "thinking"; component: Markdown | Text }> = [];
		for (let i = 0; i < message.content.length; i++) {
			const content = message.content[i];
			if (content.type === "text" && content.text.trim()) {
				// Assistant text messages with no background - trim the text
				// Set paddingY=0 to avoid extra spacing before tool executions
				const component = new Markdown(content.text.trim(), this.outputPad, 0, this.markdownTheme, undefined, {
					transform: createMarkdownTransform("assistant", this.isStreaming, this.markdownTransformers),
					streaming: this.isStreaming,
				});
				this.contentContainer.addChild(component);
				editable.push({ kind: "text", component });
			} else if (content.type === "thinking") {
				const thinkingBlocks: string[] = [];
				for (; i < message.content.length; i++) {
					const thinkingContent = message.content[i];
					if (thinkingContent.type !== "thinking") {
						break;
					}
					const thinking = thinkingContent.thinking.trim();
					if (thinking) {
						thinkingBlocks.push(thinking);
					}
				}
				i--;

				if (thinkingBlocks.length === 0) {
					continue;
				}

				// Add spacing only when another visible assistant content block follows.
				// This avoids a superfluous blank line before separately-rendered tool execution blocks.
				const hasVisibleContentAfter = message.content
					.slice(i + 1)
					.some((c) => (c.type === "text" && c.text.trim()) || (c.type === "thinking" && c.thinking.trim()));

				if (this.hideThinkingBlock) {
					// Show one static label for each run of thinking blocks when hidden.
					const component = new Text(
						theme.italic(theme.fg("thinkingText", this.hiddenThinkingLabel)),
						this.outputPad,
						0,
					);
					this.contentContainer.addChild(component);
					editable.push({ kind: "thinking", component });
				} else {
					// Render each run of thinking blocks as one Markdown section.
					const component = new Markdown(
						thinkingBlocks.join("\n\n"),
						this.outputPad,
						0,
						this.markdownTheme,
						{
							color: (text: string) => theme.fg("thinkingText", text),
							italic: true,
						},
						{
							transform: createMarkdownTransform(
								"assistant-thinking",
								this.isStreaming,
								this.markdownTransformers,
							),
							streaming: this.isStreaming,
						},
					);
					this.contentContainer.addChild(component);
					editable.push({ kind: "thinking", component });
				}
				if (hasVisibleContentAfter) {
					this.contentContainer.addChild(new Spacer(1));
				}
			}
		}
		this.streamingEditable = editable;
		this.builtWithTheme = this.markdownTheme;
		this.builtWithOutputPad = this.outputPad;
		this.builtWithHideThinkingBlock = this.hideThinkingBlock;
		this.builtWithHiddenThinkingLabel = this.hiddenThinkingLabel;
		this.builtWithTransformers = this.markdownTransformers;

		// Check if incomplete/failed - show after partial content.
		// For aborted/error tool calls, tool execution components show the error.
		// Length stops can happen before a tool call is complete, so surface them here too.
		const hasToolCalls = message.content.some((c) => c.type === "toolCall");
		this.hasToolCalls = hasToolCalls;
		if (message.stopReason === "length") {
			this.contentContainer.addChild(new Spacer(1));
			this.contentContainer.addChild(
				new Text(theme.fg("error", "Response was truncated before completion."), this.outputPad, 0),
			);
		} else if (!hasToolCalls) {
			if (message.stopReason === "aborted") {
				const abortMessage =
					message.errorMessage && message.errorMessage !== "Request was aborted"
						? message.errorMessage
						: "Operation aborted";
				this.contentContainer.addChild(new Spacer(1));
				this.contentContainer.addChild(new Text(theme.fg("error", abortMessage), this.outputPad, 0));
			} else if (message.stopReason === "error") {
				const errorMsg = message.errorMessage || "Unknown error";
				this.contentContainer.addChild(new Spacer(1));
				this.contentContainer.addChild(new Text(theme.fg("error", `Error: ${errorMsg}`), this.outputPad, 0));
			}
		}
	}

	/**
	 * Streaming fast path: when the sequence of visible text/thinking runs is
	 * unchanged between two consecutive message_update events, update the existing
	 * Markdown components in place via setText() instead of rebuilding them. This
	 * keeps the components' append-aware render cache warm, so each chunk only
	 * re-renders the appended tail instead of re-parsing the whole message.
	 */
	private tryUpdateStreamingInPlace(previous: AssistantMessage, message: AssistantMessage): boolean {
		// Rebuild if anything that affects component construction changed.
		if (
			this.markdownTheme !== this.builtWithTheme ||
			this.outputPad !== this.builtWithOutputPad ||
			this.hideThinkingBlock !== this.builtWithHideThinkingBlock ||
			this.hiddenThinkingLabel !== this.builtWithHiddenThinkingLabel ||
			this.markdownTransformers !== this.builtWithTransformers
		) {
			return false;
		}
		if (this.streamingEditable.length === 0) return false;

		const extractRuns = (msg: AssistantMessage): Array<{ kind: "text" | "thinking"; text: string }> => {
			const runs: Array<{ kind: "text" | "thinking"; text: string }> = [];
			for (let i = 0; i < msg.content.length; i++) {
				const content = msg.content[i];
				if (content.type === "text") {
					if (content.text.trim()) runs.push({ kind: "text", text: content.text });
				} else if (content.type === "thinking") {
					const thinkingBlocks: string[] = [];
					for (; i < msg.content.length; i++) {
						const thinkingContent = msg.content[i];
						if (thinkingContent.type !== "thinking") break;
						if (thinkingContent.thinking.trim()) thinkingBlocks.push(thinkingContent.thinking);
					}
					i--;
					if (thinkingBlocks.length === 0) continue;
					runs.push({ kind: "thinking", text: thinkingBlocks.join("\n\n") });
				}
			}
			return runs;
		};

		const prevRuns = extractRuns(previous);
		const nextRuns = extractRuns(message);
		if (prevRuns.length === 0 || prevRuns.length !== nextRuns.length) return false;
		for (let i = 0; i < prevRuns.length; i++) {
			if (prevRuns[i]!.kind !== nextRuns[i]!.kind) return false;
			const component = this.streamingEditable[i]?.component;
			if (!component || !("setText" in component)) return false;
		}

		for (let i = 0; i < nextRuns.length; i++) {
			const editable = this.streamingEditable[i]!;
			(editable.component as Markdown | Text).setText(nextRuns[i]!.text);
		}
		return true;
	}
}
